frontend/catalyst/compiler.py

```diff
@@ -264,233 +371,112 @@ def link(infile, outfile, flags=None, fallback_compilers=None, compile_options=N
         Raises:
             EnvironmentError: The exception is raised when no compiler succeeded.
         """
+        if outfile is None:
+            outfile = CompilerDriver.get_output_filename(infile)
         if flags is None:
-            flags = CompilerDriver._flags()
+            flags = CompilerDriver._default_flags
         if fallback_compilers is None:
             fallback_compilers = CompilerDriver._default_fallback_compilers
         # pylint: disable=redefined-outer-name
         for compiler in CompilerDriver._available_compilers(fallback_compilers):
-            success = CompilerDriver._attempt_link(
-                compiler, flags, infile, outfile, compile_options
-            )
+            success = CompilerDriver._attempt_link(compiler, flags, infile, outfile, options)
             if success:
-                return
+                return outfile
         msg = f"Unable to link {infile}. All available compiler options exhausted. Please provide a compatible compiler via $CATALYST_CC."
         raise EnvironmentError(msg)
 
 
-def lower_mhlo_to_linalg(filename: str, compile_options: Optional[CompileOptions] = None) -> str:
-    """Translate MHLO to linalg dialect.
-
-    Args:
-        filename (str): the path to a file were the program is stored.
-        Optional compile_options (CompileOptions): generic compilation options.
-    Returns:
-        a path to the output file
-    """
-    if filename[-5:] != ".mlir":
-        raise ValueError(f"Input file ({filename}) for MHLO lowering is not an MLIR file")
-
-    new_fname = filename.replace(".mlir", ".nohlo.mlir")
-
-    command = [mhlo_opt_tool]
-    command += ["--allow-unregistered-dialect"]
-    command += [filename]
-    command += mhlo_lowering_pass_pipeline
-    command += ["-o", new_fname]
-
-    run_writing_command(command, compile_options)
-
-    return new_fname
-
-
-def transform_quantum_ir(filename: str, compile_options: Optional[CompileOptions] = None) -> str:
-    """Runs quantum optimizations and transformations, as well gradient transforms, on the hybrid
-    IR.
-
-    Args:
-        filename (str): the path to a file where the program is stored.
-    Returns:
-        a path to the output file
-    """
-    if filename[-5:] != ".mlir":
-        raise ValueError(f"Input file ({filename}) for quantum transforms is not an MLIR file")
-
-    new_fname = filename.replace(".mlir", ".opt.mlir")
-
-    command = [quantum_opt_tool]
-    command += [filename]
-    command += quantum_compilation_pass_pipeline
-    command += ["-o", new_fname]
-
-    run_writing_command(command, compile_options)
+class Compiler:
+    """Compiles MLIR modules to shared objects."""
 
-    return new_fname
+    def __init__(self):
+        self.order = None
```

Comment:
The word `order` here is a reference to phase ordering, which is the order in which passes are run. Yes, will change to compilation outputs.

Reply:
Ah got it, yeah without context it's not immediately clear, especially since a dictionary is typically not thought of as storing order (though the Python one does).

Reply:
https://github.com/PennyLaneAI/catalyst/pull/38/commits/9705df55d7cd05e69f9e9c2786c9650b45647402